File: algopt/rebalancer/explorer/client/cpp/tests/ServerSelectionTest.cpp
// (c) Meta Platforms, Inc. and affiliates. Confidential and proprietary.

#include "rebalancer/explorer/client/cpp/ServerSelection.h"

#include <gtest/gtest.h>

using namespace facebook::rebalancer::explorer;
using facebook::rebalancer::explorer::client::selectBestServer;
using facebook::rebalancer::explorer::client::ServerAddr;

namespace {

ServerStatus makeStatus(int64_t loading, int64_t freeBytes) {
  ServerStatus status;
  status.loadingSandboxCount() = loading;
  status.loadedSandboxCount() = 0;
  status.freeMemoryBytes() = freeBytes;
  status.usedMemoryBytes() = 0;
  return status;
}

const std::vector<ServerAddr> kAddrs{
    ServerAddr{.host = "a", .port = 1},
    ServerAddr{.host = "b", .port = 2},
    ServerAddr{.host = "c", .port = 3},
};

bool sameAddr(const ServerAddr& lhs, const ServerAddr& rhs) {
  return lhs.host == rhs.host && lhs.port == rhs.port;
}

} // namespace

TEST(ServerSelectionTest, PicksFewestLoading) {
  const std::vector<std::optional<ServerStatus>> statuses{
      makeStatus(3, 100), makeStatus(1, 10), makeStatus(2, 100)};
  const auto best = selectBestServer(statuses, kAddrs);
  ASSERT_TRUE(best.has_value());
  EXPECT_TRUE(sameAddr(ServerAddr{.host = "b", .port = 2}, *best));
}

TEST(ServerSelectionTest, TiesBrokenByMostFreeMemory) {
  const std::vector<std::optional<ServerStatus>> statuses{
      makeStatus(1, 50), makeStatus(1, 300), makeStatus(1, 100)};
  const auto best = selectBestServer(statuses, kAddrs);
  ASSERT_TRUE(best.has_value());
  EXPECT_TRUE(sameAddr(ServerAddr{.host = "b", .port = 2}, *best));
}

TEST(ServerSelectionTest, SkipsUnreachableServers) {
  const std::vector<std::optional<ServerStatus>> statuses{
      std::nullopt, makeStatus(5, 100), std::nullopt};
  const auto best = selectBestServer(statuses, kAddrs);
  ASSERT_TRUE(best.has_value());
  EXPECT_TRUE(sameAddr(ServerAddr{.host = "b", .port = 2}, *best));
}

TEST(ServerSelectionTest, ReturnsNulloptWhenAllUnreachable) {
  const std::vector<std::optional<ServerStatus>> statuses{
      std::nullopt, std::nullopt, std::nullopt};
  EXPECT_FALSE(selectBestServer(statuses, kAddrs).has_value());
}

TEST(ServerSelectionTest, ReturnsNulloptForEmpty) {
  EXPECT_FALSE(selectBestServer({}, {}).has_value());
}

File: algopt/rebalancer/explorer/client/tests/server_selection_test.py
# (c) Meta Platforms, Inc. and affiliates. Confidential and proprietary.
# pyre-strict

import unittest

from rebalancer.explorer.client.server_selection import select_best_server, ServerAddr
from rebalancer.explorer.explorer.thrift_types import ServerStatus


def _status(loading: int, free_bytes: int) -> ServerStatus:
    return ServerStatus(
        loadingSandboxCount=loading,
        loadedSandboxCount=0,
        freeMemoryBytes=free_bytes,
        usedMemoryBytes=0,
    )


_ADDRS = [
    ServerAddr("a", 1),
    ServerAddr("b", 2),
    ServerAddr("c", 3),
]


class SelectBestServerTest(unittest.TestCase):
    def test_picks_fewest_loading(self) -> None:
        statuses = [_status(3, 100), _status(1, 10), _status(2, 100)]
        self.assertEqual(ServerAddr("b", 2), select_best_server(statuses, _ADDRS))

    def test_ties_broken_by_most_free_memory(self) -> None:
        statuses = [_status(1, 50), _status(1, 300), _status(1, 100)]
        self.assertEqual(ServerAddr("b", 2), select_best_server(statuses, _ADDRS))

    def test_skips_unreachable_servers(self) -> None:
        statuses = [None, _status(5, 100), None]
        self.assertEqual(ServerAddr("b", 2), select_best_server(statuses, _ADDRS))

    def test_returns_none_when_all_unreachable(self) -> None:
        self.assertIsNone(select_best_server([None, None, None], _ADDRS))

    def test_returns_none_for_empty(self) -> None:
        self.assertIsNone(select_best_server([], []))
